File: src/voxels/WaterClippingVolume.cpp
#include "WaterClippingVolume.h"
#include "WaterClippingUtils.h"
#include <algorithm>

WaterClippingVolume::WaterClippingVolume()
	: count(-1), isSliced(false) {
	intersectionPoints.resize(6);
}

WaterClippingVolume::~WaterClippingVolume() {
}

void WaterClippingVolume::Prepare(const WaterClippingUtils::Plane& waterClipPlane) {
	this->waterClipPlane = waterClipPlane;
	this->isSliced = WaterClippingUtils::GetCubePlaneIntersectionEdgeLoop(
		waterClipPlane, intersectionPoints, count);
}

void WaterClippingVolume::ApplyClipping(glm::vec3& vertLocalPos) const {
	if (!isSliced || waterClipPlane.GetDistanceToPoint(vertLocalPos) <= 0.0f) {
		return; // Не обрезаем, если плоскость не пересекает куб или точка ниже плоскости
	}
	
	// Проецируем точку на плоскость
	vertLocalPos = waterClipPlane.ClosestPointOnPlane(vertLocalPos);
	
	// Если точка все еще внутри куба, оставляем как есть
	if (WaterClippingUtils::IsPointInCube(vertLocalPos)) {
		return;
	}
	
	// Иначе находим ближайшую точку на цикле пересечения
	vertLocalPos = WaterClippingUtils::NearestPointOnEdgeLoop(
		vertLocalPos, intersectionPoints, count);
}

File: Water/WaterClippingVolume.cs
// Decompiled with JetBrains decompiler
// Type: WaterClippingVolume
// Assembly: Assembly-CSharp, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null
// MVID: AF8FE50B-9889-4084-9FCD-E241DDFED80F
// Assembly location: C:\Program Files (x86)\Steam\steamapps\common\7 Days To Die\7DaysToDie_Data\Managed\Assembly-CSharp.dll

using UnityEngine;

#nullable disable
public class WaterClippingVolume
{
  [PublicizedFrom(EAccessModifier.Private)]
  public Plane waterClipPlane;
  [PublicizedFrom(EAccessModifier.Private)]
  public Vector3[] intersectionPoints = new Vector3[6];
  [PublicizedFrom(EAccessModifier.Private)]
  public int count = -1;
  [PublicizedFrom(EAccessModifier.Private)]
  public bool isSliced;

  public void Prepare(Plane waterClipPlane)
  {
    this.waterClipPlane = waterClipPlane;
    this.isSliced = WaterClippingUtils.GetCubePlaneIntersectionEdgeLoop(waterClipPlane, ref this.intersectionPoints, out this.count);
  }

  public void ApplyClipping(ref Vector3 vertLocalPos)
  {
    if (!this.isSliced || (double) this.waterClipPlane.GetDistanceToPoint(vertLocalPos) <= 0.0)
      return;
    vertLocalPos = this.waterClipPlane.ClosestPointOnPlane(vertLocalPos);
    if (WaterClippingUtils.CubeBounds.Contains(vertLocalPos))
      return;
    vertLocalPos = GeometryUtils.NearestPointOnEdgeLoop(vertLocalPos, this.intersectionPoints, this.count);
  }
}
